File: sdk-java-blocking/src/test/java/dev/restate/sdk/blocking/SideEffectTest.java
package dev.restate.sdk.blocking;

import static dev.restate.sdk.core.impl.ProtoUtils.greetingRequest;

import dev.restate.sdk.core.CoreSerdes;
import dev.restate.sdk.core.impl.SideEffectTestSuite;
import dev.restate.sdk.core.impl.testservices.GreeterGrpc;
import dev.restate.sdk.core.impl.testservices.GreetingRequest;
import dev.restate.sdk.core.impl.testservices.GreetingResponse;
import io.grpc.BindableService;
import io.grpc.stub.StreamObserver;
import java.util.Objects;

public class SideEffectTest extends SideEffectTestSuite {

  private static class SideEffect extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    private final String sideEffectOutput;

    SideEffect(String sideEffectOutput) {
      this.sideEffectOutput = sideEffectOutput;
    }

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      String result = ctx.sideEffect(CoreSerdes.STRING_UTF8, () -> this.sideEffectOutput);

      responseObserver.onNext(GreetingResponse.newBuilder().setMessage("Hello " + result).build());
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService sideEffect(String sideEffectOutput) {
    return new SideEffect(sideEffectOutput);
  }

  private static class ConsecutiveSideEffect extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    private final String sideEffectOutput;

    ConsecutiveSideEffect(String sideEffectOutput) {
      this.sideEffectOutput = sideEffectOutput;
    }

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      String firstResult = ctx.sideEffect(CoreSerdes.STRING_UTF8, () -> this.sideEffectOutput);
      String secondResult = ctx.sideEffect(CoreSerdes.STRING_UTF8, firstResult::toUpperCase);

      responseObserver.onNext(
          GreetingResponse.newBuilder().setMessage("Hello " + secondResult).build());
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService consecutiveSideEffect(String sideEffectOutput) {
    return new ConsecutiveSideEffect(sideEffectOutput);
  }

  private static class CheckContextSwitching extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      String currentThread = Thread.currentThread().getName();

      String sideEffectThread =
          restateContext()
              .sideEffect(CoreSerdes.STRING_UTF8, () -> Thread.currentThread().getName());

      if (!Objects.equals(currentThread, sideEffectThread)) {
        throw new IllegalStateException(
            "Current thread and side effect thread do not match: "
                + currentThread
                + " != "
                + sideEffectThread);
      }

      responseObserver.onNext(GreetingResponse.newBuilder().setMessage("Hello").build());
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService checkContextSwitching() {
    return new CheckContextSwitching();
  }

  private static class SideEffectGuard extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();
      ctx.sideEffect(
          () -> ctx.oneWayCall(GreeterGrpc.getGreetMethod(), greetingRequest("something")));

      throw new IllegalStateException("This point should not be reached");
    }
  }

  @Override
  protected BindableService sideEffectGuard() {
    return new SideEffectGuard();
  }

  private static class SideEffectThenAwakeable extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      ctx.sideEffect(
          () -> {
            throw new IllegalStateException("This should be replayed");
          });
      ctx.awakeable(CoreSerdes.BYTES).await();

      responseObserver.onNext(GreetingResponse.newBuilder().setMessage("Hello").build());
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService sideEffectThenAwakeable() {
    return new SideEffectThenAwakeable();
  }
}

File: sdk-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/SideEffectTest.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.CoreSerdes
import dev.restate.sdk.core.impl.SideEffectTestSuite
import dev.restate.sdk.core.impl.testservices.*
import io.grpc.BindableService
import java.util.*
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.Dispatchers

class SideEffectTest : SideEffectTestSuite() {
  private class SideEffect(private val sideEffectOutput: String) :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx: RestateContext = restateContext()
      val result = ctx.sideEffect(CoreSerdes.STRING_UTF8) { sideEffectOutput }
      return greetingResponse { message = "Hello $result" }
    }
  }

  override fun sideEffect(sideEffectOutput: String): BindableService {
    return SideEffect(sideEffectOutput)
  }

  private class ConsecutiveSideEffect(private val sideEffectOutput: String) :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx: RestateContext = restateContext()
      val firstResult = ctx.sideEffect(CoreSerdes.STRING_UTF8) { sideEffectOutput }
      val secondResult =
          ctx.sideEffect(CoreSerdes.STRING_UTF8) { firstResult.uppercase(Locale.getDefault()) }
      return greetingResponse { message = "Hello $secondResult" }
    }
  }

  override fun consecutiveSideEffect(sideEffectOutput: String): BindableService {
    return ConsecutiveSideEffect(sideEffectOutput)
  }

  private class CheckContextSwitching :
      GreeterGrpcKt.GreeterCoroutineImplBase(
          Dispatchers.Unconfined + CoroutineName("CheckContextSwitchingTestCoroutine")),
      RestateCoroutineService {

    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val sideEffectThread =
          restateContext().sideEffect(CoreSerdes.STRING_UTF8) { Thread.currentThread().name }
      check(sideEffectThread.contains("CheckContextSwitchingTestCoroutine")) {
        "Side effect thread is not running within the same coroutine context of the handler method: $sideEffectThread"
      }
      return greetingResponse { message = "Hello" }
    }
  }

  override fun checkContextSwitching(): BindableService {
    return CheckContextSwitching()
  }

  private class SideEffectGuard :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = restateContext()
      ctx.sideEffect {
        ctx.oneWayCall(GreeterGrpcKt.greetMethod, greetingRequest { name = "something" })
      }
      throw IllegalStateException("This point should not be reached")
    }
  }

  override fun sideEffectGuard(): BindableService {
    return SideEffectGuard()
  }
}
